package/nmea_daemon/src/nmea_daemon.c

```diff
@@ -16,64 +16,63 @@
 #include <stdlib.h>
 #include <string.h>
 
-#include <czmq.h>
-
-#include <libpiksi/util.h>
 #include <libpiksi/logging.h>
+#include <libpiksi/endpoint.h>
+#include <libpiksi/loop.h>
 
 #define PROGRAM_NAME      "nmea_daemon"
 
 #define NMEA_PUB_ENDPOINT ">tcp://127.0.0.1:44030"  /* NMEA Pub */
 
 #define BASE_DIRECTORY    "/var/run/nmea"
 
-const char*const NMEA_GGA_OUTPUT_PATH = BASE_DIRECTORY "/GGA";
+const char* const NMEA_GGA_OUTPUT_PATH = BASE_DIRECTORY "/GGA";
 
 bool nmea_debug = false;
 
+typedef struct nmea_ctx_s {
+  pk_endpoint_t *sub_ept;
+  pk_loop_t *loop;
+  int result;
+} nmea_ctx_t;
+
+static int handle_frame_cb(const u8 *frame_data, const size_t frame_length, void *context);
+
 #define MAX_BUFFER 2048
 
 char buffer[MAX_BUFFER];
 
-static int nmea_reader_handler(zloop_t *zloop, zsock_t *zsock, void *arg)
+static void nmea_reader_handler(pk_loop_t *loop, void *handle, void *context)
 {
-  (void)zloop;
-  (void)arg;
-
-  zmsg_t *msg;
-
-  while (1) {
-    msg = zmsg_recv(zsock);
-    if (msg != NULL) {
-      /* Break on success */
-      break;
-    } else if (errno == EINTR) {
-      /* Retry if interrupted */
-      continue;
-    } else {
-      sbp_log(LOG_ERR, "error in zmsg_recv()");
-      exit(EXIT_FAILURE);
-    }
+  (void)loop;
+  (void)handle;
+  nmea_ctx_t *ctx = (nmea_ctx_t *)context;
+
+  if (pk_endpoint_receive(ctx->sub_ept, handle_frame_cb, ctx) != 0) {
+    piksi_log(LOG_ERR, "endpoint receive failed, stopping loop...");
+    ctx->result = EXIT_FAILURE;
   }
+  if (ctx->result == EXIT_FAILURE) {
```

Comment:
Can this just be moved inside the previous if?

Reply:
It can't, due to the fact that `pk_endpoint_receive` may return successfully while `handle_frame_cb` had set the result to false.